### PR TITLE
fix(802): defer string_length to Tier 3 for UTF-8 byte soundness (v0.0.182)

### LINKED ISSUE #392: Audit `smt.py` Z3 translation layer for verification soundness
The `vera/smt.py` module translates Vera contracts to Z3 assertions. A bug here could cause Z3 to verify a contract as satisfied when it isn't — silently bypassing runtime checks. This is the highest-severity correctness bug class possible in the verifier.

Audit approach:
- For every contract operator translation (arithmetic, comparison, logical), write a test with a known-false contract and verify Z3 rejects it
- Test edge cases: integer overflow, division by zero in contracts, nested quantifiers, contracts referencing `@T.result`
- Verify that Tier 1 (proved) and Tier 3 (runtime) give the same pass/fail result for the same inputs on a corpus of test functions

This is particularly important because the codebase grew rapidly — soundness gaps could have been introduced during the fast iteration phase.

### LINKED ISSUE #802: smt.py: string_length translated as Z3 code-point Length — Tier 1 proves wrong length for non-ASCII (Vera counts UTF-8 bytes)
Found by the #392 `smt.py` soundness audit. Parent: #392.

## The bug
`vera verify` proves the **wrong length** for any non-ASCII string. Z3's string theory counts Unicode **code points**; Vera's runtime `string_length` counts **UTF-8 bytes**. They disagree on every multibyte character.
```
public fn f(@String -> @Int)
  requires(@String.0 == "é")              # U+00E9: 1 code point, 2 UTF-8 bytes
  ensures(@Int.result == 1)
  effects(pure)
{ string_length(@String.0) }
```
`vera verify` → `tier1_verified: 2`, proved (Z3 `Length("é") == 1`). But the runtime returns **2** (`vera run` on `{ string_length("é") }` prints `2`), so the proved postcondition is false:
```
$ vera run f -- "é"
Error: Postcondition violation in f(@String -> @Int)
  ensures(@Int.result == 1) failed
```
For ASCII the two agree (`string_length("ab") == 2` both verifies and runs as `2`), so the bug is silent until a non-ASCII string appears.

## Root cause
`smt.py` translates `string_length` to Z3's `Length` over `z3.String` — a sequence of Unicode code points (SMT-LIB 2.6 strings). Vera strings are UTF-8 byte sequences and `string_length` returns the **byte** count. The mismatch is zero for ASCII (1 byte per code point) but ≥1 for any multibyte code point.

## Likely generalises
The same Z3-codepoint vs. Vera-byte mismatch probably affects every string builtin whose semantics are byte/offset-sensitive and modeled in Z3's string theory — `string_slice`, `string_index_of`, `string_char_code`, `string_chars`. Each is worth auditing; `string_length` is the confirmed case.

## Fix options
- Translate the byte/offset-sensitive string builtins against a **byte-indexed** model (encode lengths and indices in UTF-8 bytes) so Tier-1 reasoning matches the runtime; **or**
- Defer those builtins to Tier 3 (as numeric casts, quantifiers, and decimal arithmetic already are) rather than model them unsoundly at Tier 1.

## Regression test
The probe above must not verify `string_length("é") == 1` at Tier 1 — either prove `== 2` (byte model) or fall to Tier 3.

Refs #392 (parent).

### PR DESCRIPTION
**Pull request by aallan**
Closes #802 — the final sub-task of the [#392](https://github.com/aallan/vera/issues/392) `smt.py` soundness audit. With this, all six soundness bugs the audit found (#797–#802) are fixed, so this PR also **Closes #392**.

## The bug
`vera verify` translated `string_length` to Z3's `Length`, which counts Unicode **code points**, but Vera's runtime `string_length` counts **UTF-8 bytes**. They disagree on every multibyte character, so the verifier proved false contracts:
```
public fn f(@String -> @Int)
  requires(@String.0 == "é")      # U+00E9: 1 code point, 2 UTF-8 bytes
  ensures(@Int.result == 1)       # FALSE — runtime returns 2
  effects(pure)
{ string_length(@String.0) }
```
`vera verify` → proved at Tier 1 (Z3 `Length("é") == 1`), yet `vera run` raises a postcondition violation (`string_length("é")` is `2`). Silent for ASCII (1 byte = 1 code point); wrong for any multibyte character.

## The fix
- **`string_length`** (`vera/smt.py`): modeled at **Tier 1** only for a string **literal**, whose exact UTF-8 byte length is known (`string_length("é")` now proves `== 2`, disproves `== 1`); on any **non-literal** argument it defers to a runtime-guarded **Tier 3** obligation — Z3's string theory has no byte-length operator, so this matches the numeric-cast / quantifier / decimal precedent the audit established. Sound, and it preserves the literal-arg precision the original `z3.Length` was added for (with the *correct* byte count).
- **Audited the other modeled string builtins** and confirmed they're sound, so they're untouched: `string_contains` / `string_starts_with` / `string_ends_with` stay **Tier 1** because UTF-8 is self-synchronizing — a valid substring/prefix/suffix matches at the byte level exactly when it matches at the code-point level (verified by a verify↔run differential on non-ASCII inputs). The offset-sensitive builtins (`string_slice`, `string_index_of`, `string_char_code`, `string_chars`) were never translated to Z3 and already defer to Tier 3.

## Also found in review — astral-plane string literals (folded in)
The adversarial review of this PR surfaced a **second Z3-string unsoundness in the same class**, so I fixed it here (this PR closes the #392 audit — shipping a known Z3-string hole would make that close dishonest). Z3's string-sort alphabet only reaches U+2FFFF, and its Python binding silently stores any higher code point as the literal's *escape text* — so `string_contains("\u{10FFFF}", "f")` proved `true` at Tier 1 (matching a phantom escape byte) while the runtime returns `false`. Reproduced, then fixed in `vera/smt.py`: a string literal containing a code point > U+2FFFF defers to Tier 3 (`StringVal` returns `None`), so the predicates never match phantom bytes. `string_length` is unaffected (it byte-counts the *decoded* value, not `StringVal`). Regression test: `TestAstralStringLiteral802`. The review also prompted a **runtime** byte-length test (`string_length("é") == 2`, `"😀" == 4`) — pinning the runtime premise the verifier's byte model relies on — and escaped-unicode / 4-byte / empty-literal coverage.

## Exit checklist (§0)
- [x] `pytest tests/` — green (5,138 passed, 16 skipped, 16 deselected)
- [x] `mypy vera/` — clean
- [x] `check_conformance.py` — 93/93 (`ch04_string_builtins.vera` gains a non-ASCII byte-length case that verifies at Tier 1 and runs as `2`)
- [x] `check_examples.py` — 35/35
- [x] `check_version_sync.py` — 0.0.182
- [x] CHANGELOG `[0.0.182]` + link refs; doc-counts / site-assets / spec-examples green
- [x] New behaviour covered by tests that **fail on the pre-fix verifier** — `tests/test_string_length_soundness.py` (RED-confirmed: pre-fix proves `== 1` / disproves `== 2`)

## Tests
- **New** `tests/test_string_length_soundness.py` (8): the literal byte-model (`"é"` → `2`, not `1`), slot-arg deferral (the issue's probe no longer proves at Tier 1), ASCII preservation, and the predicate-soundness guards (verify↔run agree on non-ASCII).
- **Rewritten** tests that *pinned the old unsound behaviour* — `TestStringLengthVerification` asserted `string_length` was Tier-1 via `z3.Length` (it now asserts the deferral / literal byte-model). Refinement, E501-rendering, and #667 call-typing tests that used `string_length` only as an *example* builtin switch to `string_starts_with` (a still-Tier-1 predicate), preserving each test's actual purpose. The aggregate tier-count test moves -2 T1 / +2 T3 (two example contracts defer), total unchanged.

## Reproduce-before-fix note (CR's suggested option)
The issue offered a byte-indexed Z3 model as the alternative to deferral. A *full* byte model would require re-encoding Vera strings as Z3 byte-sequences (`SeqSort` of `BitVec8`) — a large rework that would itself need an audit. The literal-byte-length model here fixes the unsoundness, keeps the documented literal-arg precision, and minimises collateral weakening (sound checks like `string_length("") > 0` over a literal still decide statically); a richer byte model can follow if Tier-1 string-length reasoning over variables is ever needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved verifier soundness for `string_length` by modelling supported string literals using exact UTF-8 byte length; non-literal inputs now defer to runtime-guarded Tier 3 checks, with added safeguards for astral and lone-surrogate edge cases.
  * Retained Tier 1 handling for `string_contains`, `string_starts_with`, and `string_ends_with` where sound, with Tier 3 deferrals for problematic literals.
* **Tests**
  * Added a new `string_length` soundness regression suite and updated tiering/verification expectations, including an end-to-end contract-violation case for issue 802.
* **Documentation**
  * Updated spec chapter 6.4.3 and release materials for `v0.0.182`, refreshing documented release and test-status totals.

### CODE REVIEW

File: tests/test_string_length_soundness.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    def test_four_byte_emoji_literal_byte_length(self) -> None:
+        # A 4-byte UTF-8 character (U+1F600) — code-point count 1, byte count 4,
+        # the case where code points and bytes diverge most.
+        ok = _verify("""
+public fn f(@Unit -> @Int)
+  requires(true) ensures(@Int.result == 4) effects(pure)
+{ string_length("\\u{1F600}") }
+""")
+        assert _ok(ok) and ok.summary.tier1_verified >= 1
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_string_length_soundness.py:132]: @coderabbitai Added a boundary test in `2961d91` — but on a *predicate*, not `string_length`. `string_length` byte-counts via Python (`len(value.encode("utf-8"))`), so `string_length("\u{2FFFF}")` and `string_length("\u{30000}")` both verify `== 4` and never touch the `z3.StringVal` alphabet cutoff. `test_astral_cutoff_boundary` pins it via `string_starts_with`: `\u{2FFFF}` stays Tier 1 (`tier3_runtime == 0`), `\u{30000}` (one above) defers (`tier3_runtime >= 1`) — catching a `>` vs `>=` regression.

File: tests/test_string_length_soundness.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        result = _verify("""
+public fn check(@String -> @Bool)
+  requires(@String.0 == "\\u{10FFFF}")
+  ensures(@Bool.result == true)
+  effects(pure)
+{ string_contains(@String.0, "f") }
+""")
+        assert result.summary.tier1_verified == 0, (
+            result.summary.tier1_verified, result.summary.tier3_runtime,
+        )
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_string_length_soundness.py]: @coderabbitai Done in `2961d91` — reworked the test to the literal-on-literal form (`string_contains("\u{10FFFF}", "f")`) and it now asserts `_ok(result)` and `tier3_runtime >= 1`, the stronger pins you asked for. One note on why that form: an astral literal in a `requires` is *dropped* as untranslatable, which leaves the false `ensures` unprovable and yields a loud E500 — so `_ok` wouldn't hold there. The literal-on-literal form is the direct unsoundness and defers cleanly to Tier 3, which is exactly the fallback worth pinning.

**Comment by aallan** [comment]
@coderabbitai Good catch — reproduced (it crashed `vera verify` with `UnicodeEncodeError`) and fixed in `2961d91`. A lone surrogate broke *both* paths: `string_length`'s `value.encode("utf-8")` (now wrapped → Tier 3 on `UnicodeEncodeError`) and `z3.StringVal` in the predicate path (the literal-deferral guard now covers U+D800–U+DFFF alongside the `> U+2FFFF` astral cutoff). Regression: `test_surrogate_literal_does_not_crash` exercises both a `string_length` and a `string_contains` over `"\u{D800}"`.

File: spec/06-contracts.md [hunk 3]
```diff
@@ -254,6 +254,8 @@ The `@Nat` obligations (E502 / E503) carry the most nuance, spanning many bindin
...
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/06-contracts.md]: @coderabbitai Fixed in `7376df2` — reworded so the U+2FFFF cutoff is scoped to the `z3.StringVal`-based predicate translation (`string_contains` / `string_starts_with` / `string_ends_with`). `string_length` byte-counts the *decoded* literal, so an astral (> U+2FFFF) literal's length stays Tier 1; only a lone surrogate (no UTF-8 encoding) defers `string_length` too, on both paths.

File: tests/test_string_length_soundness.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+  requires(true) ensures(@Int.result == 1) effects(pure)
+{ string_length("\\u{D800}") }
+""")
+        assert length.summary.tier3_runtime >= 1  # deferred, no crash
+        predicate = _verify("""
+public fn f(@Unit -> @Bool)
+  requires(true) ensures(@Bool.result == true) effects(pure)
+{ string_contains("\\u{D800}", "x") }
+""")
+        assert predicate.summary.tier3_runtime >= 1  # deferred, no crash
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_string_length_soundness.py]: @coderabbitai Done in `7376df2` — `test_surrogate_literal_does_not_crash` now asserts `_ok(length)` and `_ok(predicate)` alongside the `tier3_runtime >= 1` checks, so it fails on an E500 / checker error rather than accepting any deferral. Both literal-on-literal probes verify cleanly to Tier 3, so `_ok` holds.

**Comment by aallan** [comment]
## pr-review-toolkit review (4 agents, parallel)

Ran a comprehensive review of this PR — **code-reviewer, pr-test-analyzer, silent-failure-hunter, comment-analyzer** — against head `b350578`.

**Critical / Important: none.** All four agents independently validated the fix as sound:
- The literal byte-model + non-literal Tier-3 deferral is conservative (`None` → Tier 3 verified at every `translate_expr` consumer — never a false discharge); the `except UnicodeEncodeError` is correctly narrow.
- The astral cutoff (`> U+2FFFF`) and surrogate guards are complete w.r.t. the inputs Z3 can't model; **mutation-validated** (reverting the byte-model / flipping `>` to `>=` / removing the surrogate guard each flips the right tests RED).
- The UTF-8 self-synchronization soundness of the kept predicates was re-checked over ~20k randomized cases (0 mismatches).

**Addressed (the only findings — doc accuracy + one test), in `b350578`:**
- Two agents flagged that the comments/spec/CHANGELOG/test-docstring said `z3.StringVal` *raises* on a lone surrogate. Verified on z3 4.16: it does **not** — it silently stores the code point's escape text (same as the astral case); only `string_length`'s `encode("utf-8")` genuinely raises. The guard returns `None` *before* `StringVal` either way, so behaviour was always sound — only the rationale was wrong. Reworded `vera/smt.py`, `spec/06`, `CHANGELOG`, the surrogate test docstring; corrected the astral escape-text illustration to the actual `\u{...}` form.
- Added a runtime **soundness-loop** test (`tests/test_codegen.py`): a slot-arg `string_length` contract that `vera verify` defers to Tier 3 raises a postcondition violation at runtime — closing the "deferral is sound, not just imprecise" loop the verifier-side tests leave open.

No behaviour change in this round.

File: tests/test_codegen.py [hunk 5]
```diff
@@ -1203,6 +1203,45 @@ def test_string_length_with_escapes(self) -> None:
...
+        source = r'''
+public fn f(@String -> @Int)
+  requires(true) ensures(@Int.result == 1) effects(pure)
+{ string_length(@String.0) }
+'''
+        result = _compile_ok(source)
+        with pytest.raises(
+            (wasmtime.WasmtimeError, wasmtime.Trap, RuntimeError)
+        ):
+            execute(result, fn_name="f", raw_args=["é"])
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen.py:1248]: @coderabbitai Valid — fixed in 2d2db6e. Confirmed empirically that the deferred postcondition raises `WasmTrapError` with `kind == "contract_violation"` (message "Postcondition violation"), so the broad `raises((WasmtimeError, Trap, RuntimeError))` could green-pass on an unrelated failure. Narrowed to `raises(WasmTrapError)` + `assert excinfo.value.kind == "contract_violation"`; mutation-checked that asserting any other kind flips the test RED.

**Comment by aallan** [comment]
@coderabbitai Re: the outside-diff finding on `tests/test_string_length_soundness.py` `test_astral_cutoff_boundary` (217-236) — valid on a precision point, fixed in 2d2db6e. The mechanism is slightly different from the writeup: the slot-arg `requires(@String.0 == "\u{30000}")` form produces *two* tier3 obligations (the astral requires literal defers as an obligation, not just an assumption), so `tier3 >= 1` couldn't isolate the body-predicate cutoff — a body-only regression to Tier 1 would still pass. Rewrote both sides to literal-on-literal with `requires(true)`, so a single `ensures` obligation is driven solely by the body predicate's astral literal, and added `assert _ok(deferred)`. Mutation-checked: removing the `> 0x2FFFF` guard now flips the isolated `tier3 >= 1` assertion RED.

**Comment by aallan** [comment]
@coderabbitai resume

**Comment by aallan** [comment]
@coderabbitai full review

File: TESTING.md [hunk 6]
```diff
@@ -75,14 +75,15 @@ python scripts/fix_allowlists.py --fix               # auto-fix stale allowlists
...
 | `test_stress.py` | 16 | 553 | Scale-dependent regression tests (#596) — `@pytest.mark.stress`, skipped by default.  9 logical tests × eager-GC lane parametrisation = 16 test instances.  10K `array_map`, 5K nested-array `array_map`, 1K-deep tail recursion with allocating arg, 1M-deep tail recursion with allocating arg (#549 GC-aware TCO), 20×20 nested array-fold-of-array-fold, 100K `array_fold`, 10K String allocations, 1K `State<Int>` get/put cycles, 10K `IO.print` calls.  Pins #570 / #515 / #593 / #549 / #487 / #348 / #573 regression coverage |
+| `test_string_length_soundness.py` | 15 | 278 | #802 — string_length code-point vs UTF-8 byte soundness: a non-literal `string_length` defers to Tier 3 (the issue's `"é"` probe no longer proves `== 1` at Tier 1), a string-literal length is modeled at its exact UTF-8 byte count (`== 2` for `"é"`), and the boolean predicates `string_contains` / `string_starts_with` / `string_ends_with` stay Tier 1 (sound under UTF-8 self-synchronization) |
```
[the remark this answers is not in the document]
Comment (aallan) [reply TESTING.md]: @coderabbitai Valid — fixed in 2f4ba0b. The row did omit the astral/surrogate literal predicate coverage (`TestAstralStringLiteral802`). Appended a clause: a predicate over an astral (> U+2FFFF) or lone-surrogate literal defers to Tier 3 (z3.StringVal cannot model those code points), so the row now matches the file's coverage.

**Comment by aallan** [comment]
@coderabbitai full review
